CIP-????/README.md

```diff
@@ -0,0 +1,530 @@
+   ; 2) every transaction_index must be strictly smaller than the
+   ;    length of transaction_bodies
+```
+
+Just like with the transaction cddl, the block's transaction bodies and the transaction witness sets
+will be merged, and the constraints must be updated accordingly:
+
+```cddl
+; This is re-posted here for reference.
+transaction_piece =
+  { 0: transaction_body
+  , 1: transaction_witness_set ; The witness set for this transaction body.
+  }
+
+; A specific transaction in the block is a specific batch of transaction pieces. 
+block_transaction = [* transaction_piece]
+
+block =
+  [ header
+  , block_transactions   : [* block_transaction] ; This is new
+  , auxiliary_data_set   : {* transaction_index => auxiliary_data }
+  , invalid_transactions : [* transaction_index ]
+  ]; Valid blocks must also satisfy the following constraint:
+   ; 1) every transaction_index must be strictly smaller than the
+   ;    length of block_transactions
+```
+
+### Transaction Size Limit
+
+A transaction broken into pieces should still have approximately the same total size as the whole
+transaction. There are a few things to note:
+
+- If two transaction pieces require the same witness (ie, both pieces require a signature from
+Alice), this witness would now effectively appear twice.
+- Certain items can appear in multiple pieces (eg, the same input can be spent/referenced, tokens can
+be minted/burned that would otherwise cancel out, etc).
+
+Despite the above, it is probably fine to leave the maximum transaction size as is, at least to
+start. The transaction fee would naturally incentivize users/aggregators to minimize the amount of
+duplication in a transaction.
+
+### Block Size Limit
+
+Due to the possible duplicate items in each transaction and the fact that a block is made up of
+multiple transactions, the issues in the Transaction Size Limit section are slightly exacerbated.
+However, it should still be fine to keep the current block size limit, at least initially. Just like
+with transactions, the transaction fee would naturally incentivize users to minimize duplications.
+
+### Smart Contract Execution
+
+While the cryptographic keys only need to sign the relevant transaction pieces, the smart contracts
+would be executed against the full transaction. This requirement enables the current smart contract
+interface to remain mostly untouched. The flow is essentially:
+
+1. Deterministically derive the whole transaction from the list of pieces.
+2. Generate the script context for the whole transaction.
+3. Evaluate the smart contract using this script context.
+
+When combining the transaction pieces, the required executions (ie, script + datum + redeemer) can
+be built up as well. Duplicate executions should be removed. For example, if two transaction pieces
+look to spend the same smart contract input using the same redeemer, there is no reason to evaluate
+this scenario twice since they are guaranteed to have the same result.
+
```

Comment:
This looks like a very promising design indeed! I am curious to understand how minting NFTs that require the spending of a specific UTxO entry would work, if two pieces are trying to mint the same NFT? Would you be able to locally test if your NFT contract validates?

Reply:
> I am curious to understand how minting NFTs that require the spending of a specific UTxO entry would work, if two pieces are trying to mint the same NFT? Would you be able to locally test if your NFT contract validates?

Hmm, I hadn't thought about testing just a piece, but it should be possible since a transaction piece is just a transaction body. This would just require a smart contract to be able to run against an unbalanced transaction body. Whether the smart contract would still succeed when aggregated with other pieces depends on the smart contract logic. So Alice can verify her piece will successfully validate *before* she signs, and the aggregator can verify it will still validate *after* adding it to the batch.

Reply:
In general, it opens up the possibility of an attack if the outcome of contract validation can be changed after you submit your transaction or transaction piece. I see two ways to deal with this :

1. restrict your design to only allow combining pieces that contain no Plutus smart contracts whatsoever

2. simplify Validation Zones to be more in line with what you are suggesting/be more implicit with value flow :
- revert changes to the UTxO rule, UTxO State, and Transaction data structures (leave them the same as in the current non-zone ledger design)
- remove new Plutus version requirement
- move the `produced = consumed` computation to the ZONE rule, and aggregate across all transactions in the zone, that is `Sum_{tx \in zone} consumed tx = Sum_{tx \in zone} produced tx`
- replace the cycle detection in the ZONE rule by in independency check, which checks that "for all inputs `(txid, ix)` in all transactions in a zone, `txid` is not the ID of another transaction in the zone
- only require the per-transaction component of the transaction fee to be paid once per zone

The validation zones simplification will all remove the need for the weird `requireTxs` song and dance, and an aggregator would be able to spend their liquidity pool once per zone to provide any missing value and claim any extra value within a single transaction. Scripts would operate without any changes. Flash loan attacks would be prevented by the independency check.

Reply:
1. The difference between an account-based dApp doing things and the Boolean outcome of an EUTxO contract is that the dApp can (and has to) be written in a way that accommodates any possible account state to which it can be applied, eg price changes etc. An EUTxO contract fee cannot even be calculated without the complete fixed transaction data.  You can make any contract fail by bulking up the tx size. If you make a transaction that is valid, someone can always make your scripts fail by submitting it batched with whatever nonsense. To avoid this, you have to be able to prevent your transaction from being modified (this is currently possible only by signing it with a key that is required to sign it). Note that even if you yourself didn’t include collateral, and are relying on someone else to include it, the transaction batch they construct can also be messed up. This brings me to the Validation Zones modification - 

2. The change I propose actually allows you to model unbalanced intents by allowing whole transactions that are unbalanced, while the zones containing transactions must be balanced. It is somewhere in between the original design and what you propose. Instead of requests and fulfills, it simply allows assets to be missing from or be extra in a transaction. The preservation of value at the zone level ensures that regardless of who provides the missing assets and claims the extra ones , things balance out. Consider zone [Tx1; Tx2]

Tx1 - 
(Resolves) inputs: txIn -> (10 Ada, …)
Outputs:(5PolinaCoin, …)

Tx2 - 
(Resolved) inputs :txIn’ -> (5PolinaCoin, …)
Outputs :(10Ada, …)

The important thing is that txIn’ is not hash (Tx1), because of flash loans. No need to request 5 PolinaCoin or make an unlocked offer of 10Ada.

Reply:
The weakening to transaction balancing seems like a good middle-of-the-road approach. It means you can optionally have interdependencies but aren't forced to exactly specify everything. I suspect that in practice all it really adds is a minor optimization on transaction size, but that's nice as well.

Reply:
I have a few questions:

- Transaction Fees - is it possible for unbalanced transactions within a zone to have a fee of zero? Particularly for babel fees, only one transaction in the zone needs to cover the fee for the entire zone. Multiple transactions in the zone could specify the fee (to pay their fair share) in which case the total fee paid would be the sum of all of the fees paid in the zone.
- Flash loans - I always thought flash loans implied no time passing; you borrow a loan and immediately pay it back "before the world notices". But that doesn't seem to be possible here since time is implied to pass: tx1 is processed before tx2. How is this any different than what currently happens where one transaction is processed after another in the same block? The only issue I can see is that the transaction fee would be lower now (assuming it is 0.15 ADA per zone now instead of 0.15 ADA per transaction). I don't see why `txIn'` can't be `= (hash (Tx1), _)`. If it could, it seems trivial to do an atomic batch of transactions since they are intrinsically linked.

Reply:
- i think it will not be possible to specify a fee of 0 (this is because the amount of fee specified in each transaction must go to the fee pot, and that's how the system knows how much to add), but it does not mean you have to provide the fee. Your transaction can simply be short that amount, and someone else's transaction needs to have that amount extra to make a valid zone

- the flash loan problem here really has to do with smart contracts : you should not be able to use a thread token or a role token in your transaction unless you spend an actual output containing one. If we allow transactions in a zone to be chained in this way, you dont actually need to spend a real output with a role token to use one, flash-loaning yourself one :

Tx1 -
(Resolved) inputs : txIn -> (out1)
Outputs: (1RoleToken, …) ,  (out2)

where out1 and out2 are some outputs containing the state of some contract which can only be updated if the transaction contains a RoleToken that allows such an update to be made

Tx2 -
(Resolved) inputs : (hash (Tx1), 1) -> (1RoleToken, …)

Reply:
> you should not be able to use a thread token or a role token in your transaction unless you spend an actual output containing one.

Perhaps I misunderstood what you meant by `(hash (Tx1), _)`. I may also be misunderstanding what you mean by
> (Resolved) inputs: txIn -> (out1)`. 

The way I am thinking about it, the role token requirement is guaranteed to be true; there is no way to NOT spend the role token. No "resolving" is required.

A zone is just a list of transaction: `[tx1, tx2, tx3]`. But we can force a dimension of time on the list:
```
---- Time ---->
[tx1, tx2, tx3]
```

If we do this, then the above is just a mini-blockchain, and we can treat it as such. The zone can build up its own mini UTxO set, and transactions in the zone have access to the current state of `real UTxO set + zone UTxO set` at the time they are being processed. The zone can keep track of both the UTxOs spent from the real UTxO set and the UTxOs created/spent from the zone UTxO set. When all transactions in the zone have been fully validated, the net affect of the real UTxO set + the zone UTxO set is what actually gets applied to the blockchain.

In the above zone, `tx1` gets processed first. Imagine if `tx1` spends a real output `r1` and produces an output `z1`. The zone would log that `r1` has been spent from the real UTxO set and `z1` has been added to the zone UTxO set. `tx2` comes after `tx1` and can spend any remaining UTxO in the real UTxO set and the zone UTxO set. So `tx2` cannot spend `r1` because it was already spent by `tx1`, but it can spend both `r2` (another input from the real set) and `z1` to produce `z2`. `tx2` would call `z1` explicitly by output reference: `(hash(tx1), z1_index)`. `z1_index` is its index in `tx1`'s output list.

Now the new state is the real UTxO set minus `r1` and `r2`, plus the zone UTxO set of `z2`. `tx3` comes after `tx2` so it must use this new state. If it wants to spend `z2`, it must call it explicitly by output reference: `(hash(tx2), z2_index)`. `z2_index` is its index in `tx2`'s output list. Let's say `tx3` spends `z2` and produces `z3`. `tx3` cannot spend `z1` because it was already spent by `tx2`.

The net effect of this zone is `r1` and `r2` are both spent to produce `z3`. This is the change that would actually be recorded on the blockchain. 

With this method, if the role token was in `r1`, `tx1` moved it to `z1`. If `tx2` wants to execute the smart contract that requires the role token, it must now spend `z1` because that is where the token is at the time `tx2` is being processed. If `tx3` wants to use the role token as well, it must spend `z2` since that is where `tx2` moved the token to. Because time is enforced on the zone, the zone can know where the role token is at all times. When `z3` is actually recorded on-chain, it is the result of `r1` -> `z1` -> `z2` -> `z3`. The eUTxO model guarantees the result is the same as doing the updates separately (ie, in separate zones).

> [!Important]
> With this approach, building a transaction chain **requires** coordination between the transactions in the chain. However, I am making the assertion that: *coordination is always possible whenever transaction chaining is useful*.  If we accept this assertion, there is no need for any "resolving", and flash loans are impossible.

Strangers can't build up chains this way since they don't know each others' transactions (and therefore each others' output references), but why would strangers be chaining their transactions anyway? This is mainly useful for account style DApps that are controlled by batchers.

A batcher can personally build up the chain when trying to fulfill the user orders. They can create `tx1` first and use its hash in `tx2`, and then use `tx2`'s hash in `tx3`. Furthermore, since the batcher knows the order of the outputs in each transaction, they know what the resulting output references will be in advance. The batcher can even turn the user orders into chains (if the user transactions allow for it). If batchers need to build up a chain together, they can coordinate this off-chain.

Because the outputs are being called explicitly by output reference, it is trivial to check if the role token is actually being spent. The smart contracts don't even need to know they are being executed in a zone.

Reply:
> i think it will not be possible to specify a fee of 0 (this is because the amount of fee specified in each transaction must go to the fee pot, and that's how the system knows how much to add)

I don't see why it is possible for the zone to know if it is balanced, but not know if it has enough ADA provided for the fee. The zone checks should have everything they need to calculate the total required fee for the zone, and it can compare this against the sum of the fee specified in each transaction within the zone. This would enable some transactions to have a fee of zero while still being part of a valid zone.

Reply:
> I don't see why it is possible for the zone to know if it is balanced, but not know if it has enough ADA provided for the fee. 

This is maybe possible? Removing the per-transaction fee check is likely possible, and if we do, entering a paid fee of 0 for a transaction would also be possible. I need to think more about this, though, and would be happy to discuss in CIP meeting.

I think i need a diagram for your previous comment ...

Here is a draft of the "Implicit Value Flow" validation zones version, it is much simpler than the original

https://github.com/cardano-foundation/CIPs/pull/862/commits/1fd1d6dc56fadca8b07009e795b3ec4eedc298cd#diff-560051716ec259fa49b3ef402bb6c1f4b7517d1bf400db3e735996986398bf96

Reply:
@AndrewWestberg @WhatisRT  if you feel like having a look also